pkg/network/proxy/proxy.go

```diff
@@ -161,35 +177,34 @@ func (proxy *OsdnProxy) handleAddOrUpdateNetNamespace(obj, _ interface{}, eventT
 	netns := obj.(*networkv1.NetNamespace)
 	klog.V(5).Infof("Watch %s event for NetNamespace %q", eventType, netns.Name)
 
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
-	proxy.ids[netns.Name] = netns.NetID
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	ns := proxy.getNamespace(netns.Name)
+	ns.global = (netns.NetID == network.GlobalVNID)
 }
 
 func (proxy *OsdnProxy) handleDeleteNetNamespace(obj interface{}) {
 	netns := obj.(*networkv1.NetNamespace)
 	klog.V(5).Infof("Watch %s event for NetNamespace %q", watch.Deleted, netns.Name)
 
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
-	delete(proxy.ids, netns.Name)
-}
-
-func (proxy *OsdnProxy) isNamespaceGlobal(ns string) bool {
-	proxy.idLock.Lock()
-	defer proxy.idLock.Unlock()
+	proxy.Lock()
+	defer proxy.Unlock()
 
-	if proxy.ids[ns] == network.GlobalVNID {
-		return true
+	ns := proxy.namespaces[netns.Name]
+	if ns == nil {
+		return
 	}
-	return false
+
+	ns.global = false
+	proxy.maybeDeleteNamespace(netns.Name, ns)
 }
 
 func (proxy *OsdnProxy) updateEgressNetworkPolicy(policy networkv1.EgressNetworkPolicy) {
-	ns := policy.Namespace
-	if proxy.isNamespaceGlobal(ns) {
+	ns := proxy.getNamespace(policy.Namespace)
```

Comment:
this call is not under the `proxy.Lock()`

Reply:
It is. `updateEgressNetworkPolicy` requires its caller to have gotten the lock. I added some comments clarify this.